### PR TITLE
Check input dimensions in several models

### PR DESCRIPTION
**Pull request by abelsiqueira**
Very large update to check whether input vectors have the right dimensions.
FeasibilityResidual is not included.

### CODE REVIEW

**Comment by abelsiqueira** [comment]
:robot: Testing breakage of this pull request

| Package Name | master | stable |
|--|--|--|
| AmplNLReader | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/671075742) | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/671075743) | 
| CUTEst | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/671075744) | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/671075745) | 
| CaNNOLeS | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/671075746) | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/671075747) | 
| NLPModelsIpopt | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/671075748) | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/671075749) | 
| NLPModelsJuMP | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/671075750) | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/671075752) | 
| QuadraticModels | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/671075754) | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/671075757) | 
| SolverTools | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/671075758) | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/671075760) |

**Comment by abelsiqueira** [comment]
:robot: Testing breakage of this pull request

| Package Name | master | stable |
|--|--|--|
| AmplNLReader | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/671089823) | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/671089824) | 
| CUTEst | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/671089825) | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/671089826) | 
| CaNNOLeS | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/671089827) | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/671089828) | 
| NLPModelsIpopt | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/671089829) | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/671089830) | 
| NLPModelsJuMP | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/671089831) | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/671089832) | 
| QuadraticModels | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/671089833) | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/671089834) | 
| SolverTools | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/671089835) | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/671089836) |

**Comment by dpo** [comment]
We could do the same in basically all of our packages. Should it be factored out into a separate package?

**Comment by abelsiqueira** [comment]
It's separated in the tests, like `consistency` and `multiple_precision`. It can be used in the other packages in the same way, i.e., include file `NLPModels.jl/test/check-dimensions.jl` and call `check_nlp_dimensions` and `check_nls_dimensions`.

**Comment by dpo** [comment]
That's right for packages that use NLPModels, but it could also be used in, e.g., LinearOperators. Not a big deal. It's easy to add a macro there and will avoid an extra dependency.

**Comment by dpo** [comment]
By the way, we already had https://github.com/JuliaSmoothOptimizers/NLPModels.jl/blob/v0.11.1/src/nlp_utils.jl

**Comment by dpo** [comment]
I like that you added an exception here. Should we combine it with the existing `@lencheck`?

**Comment by abelsiqueira** [comment]
I'll take a look on this now.

**Comment by dpo** [comment]
Ok. We really should use something like this across the board.

File: src/NLPModels.jl [hunk 1]
```diff
@@ -29,22 +29,22 @@ export reset_data!, reset!, sum_counters,
        hess_structure!, hess_structure, hess_coord!, hess_coord, hess, hprod, hprod!, hess_op, hess_op!,
        push!,
        varscale, lagscale, conscale,
-       NotImplementedError
+       NotImplementedError, DimensionError, @check_dim
```
Comment (dpo) [inline src/NLPModels.jl:32]: Still need `@check_dim`?

Comment (abelsiqueira) [reply src/NLPModels.jl]: No, fixed it.

**Comment by abelsiqueira** [comment]
:robot: Testing breakage of this pull request

| Package Name | master | stable |
|--|--|--|
| AmplNLReader | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/674429124) | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/674429125) | 
| CUTEst | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/674429126) | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/674429127) | 
| CaNNOLeS | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/674429128) | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/674429129) | 
| NLPModelsIpopt | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/674429130) | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/674429131) | 
| NLPModelsJuMP | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/674429132) | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/674429133) | 
| QuadraticModels | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/674429134) | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/674429135) | 
| SolverTools | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/674429136) | [](https://travis-ci.org/JuliaSmoothOptimizers/NLPModels.jl/jobs/674429137) |

**Comment by dpo** [comment]
Thank you!

**Pull request merged**